internal/executor/instance/pipe.go

```diff
@@ -0,0 +1,78 @@
+package instance
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/cirruslabs/cirrus-ci-agent/api"
+)
+
+var ErrPipeCreationFailed = errors.New("failed to create pipe instance")
+
+type PipeStage struct {
+	Image                  string
+	CommandFrom, CommandTo string
+}
+
+type PipeInstance struct {
+	Stages []PipeStage
+	CPU    float32
+	Memory uint32
+}
+
+// PipeStagesFromCommands uses image hints in commands to build the stages.
+func PipeStagesFromCommands(commands []*api.Command) ([]PipeStage, error) {
+	var stages []PipeStage
+
+	for i, command := range commands {
+		image, found := command.Properties["image"]
+		if !found {
+			if i == 0 {
+				return nil, fmt.Errorf("%w: first command does not have an image property", ErrPipeCreationFailed)
+			}
+
+			continue
+		}
+
+		// Close old stage
+		if len(stages) != 0 {
+			stages[len(stages)-1].CommandTo = command.Name
+		}
+
+		// Open new stage
+		stages = append(stages, PipeStage{
+			Image:       image,
+			CommandFrom: command.Name,
+		})
+	}
+
+	return stages, nil
+}
+
+func (pi *PipeInstance) Run(ctx context.Context, config *RunConfig) error {
+	workingVolume, err := CreateWorkingVolume(ctx, config.ProjectDir)
+	if err != nil {
+		return err
+	}
+
+	for _, stage := range pi.Stages {
+		params := &Params{
+			Image:             stage.Image,
+			CPU:               pi.CPU,
+			Memory:            pi.Memory,
+			CommandFrom:       stage.CommandFrom,
+			CommandTo:         stage.CommandTo,
+			WorkingVolumeName: workingVolume.Name(),
+		}
+
+		if err := RunDockerizedAgent(ctx, config, params); err != nil {
```

Comment:
What will happen here if the first stage fails? Seems there will be no error and a pipe will proceed, wouldn't it?

Reply:
Fixed via https://github.com/cirruslabs/cirrus-cli/pull/26/commits/772c9e655f1d4600e40201ace81c88eac46ded9e.